pkg/nimble: add static connection manager `statconn`

haukepetersen opened the pull request:
### Contribution description
This PR adds a new connection manager for `nimble_netif` that can manage connections statically. The main point of this module is, that connections are actively maintained, so that if two nodes get disconnected for whatever reason, that `statconn` module takes care of re-opening that connection as soon as possible.

This new module is especially helpful for long-running experiments that are build upon user-defined topologies, but it may also be useful in other contexts.

### Testing procedure
Flash the included `tests/nimble_statconn_gnrc` example on any two or more `nimble_netif`-capable boards of your choice. Open a connection by using the included `statconn` shell command:

Node A:
```
statconn adds ADDR_OF_NODE_B
```
Node B:
```
statconn addm ADDR_OF_NODE_A
```
The nodes should now open a BLE connection between them and it should be possible to ping them using their link local IPv6 addresses. Further the `ble info` shell command should show the connection.

Now use the `ble close X` shell command on either node to manually drop the connection. The connection should be (more or less) instantly be re-opened by the `statconn` module.

### Issues/PRs references
none

haukepetersen commented:
Anyone able to take a look at this PR? Thx!

miri64 commented:
@haukepetersen needs rebase. @fjmolinas @aabadie could maybe one of you test this?

haukepetersen commented:
rebased

miri64 commented:
> @fjmolinas @aabadie could maybe one of you test this?

Ping? I would use the testbed, if the nodes were free.

miri64 commented:
(can't flash my nodes here locally since I am missing the programmer ^^")

haukepetersen commented:
> if the nodes were free.

Sorry, I am occupying most of them at the moment :-)

But nodes `nrf52840dk-1-5` should be available, right? They should do the trick...

miri64 commented:
> But nodes `nrf52840dk-1-5` should be available, right? They should do the trick...

Ahhh sorry, I somehow scrolled over them. Sure! Will use them :-)

miri64 commented:
I think this PR needs another rebase and adaptation to #14289 :(

```
make: Entering directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/nimble_statconn_gnrc'
Building application "tests_nimble_statconn_gnrc" for "nrf52840dk" with MCU "nrf52".

[INFO] cloning nimble
mkdir: cannot create directory ‘/home/mlenders/Repositories/RIOT-OS/RIOT/build/pkg’: Permission denied
make[1]: *** [/home/mlenders/Repositories/RIOT-OS/RIOT/pkg/pkg.mk:122: /home/mlenders/Repositories/RIOT-OS/RIOT/build/pkg/nimble/.git] Error 1
make: [/home/mlenders/Repositories/RIOT-OS/RIOT/tests/nimble_statconn_gnrc/../../Makefile.include:670: pkg-prepare] Error 2 (ignored)
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/pkg/nimble
[INFO] cloning nimble
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/pkg/nimble/addr/
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/build/pkg/nimble/nimble/controller/src/ -f /home/mlenders/Repositories/RIOT-OS/RIOT/Makefile.base MODULE=nimble_controller
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/build/pkg/nimble/nimble/drivers/nrf52/src/ -f /home/mlenders/Repositories/RIOT-OS/RIOT/Makefile.base MODULE=nimble_drivers_nrf5x
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/build/pkg/nimble/nimble/host/src/ -f /home/mlenders/Repositories/RIOT-OS/RIOT/pkg/nimble/nimble.host.mk
mkdir: cannot create directory ‘/home/mlenders/Repositories/RIOT-OS/RIOT/build/pkg’: Permission denied
make[1]: *** [/home/mlenders/Repositories/RIOT-OS/RIOT/pkg/pkg.mk:122: /home/mlenders/Repositories/RIOT-OS/RIOT/build/pkg/nimble/.git] Error 1
```

miri64 commented:
(maybe not even adaptation as this is a nimble problem and nimble compiles in master)

miri64 commented:
Ok no, sorry local issue. `build/` somehow was set with root permissions.

aabadie commented on pkg/nimble/Makefile:
```diff
@@ -88,3 +88,6 @@ nimble_scanlist:
 	"$(MAKE)" -C $(TDIR)/scanner
+
+nimble_statconn:
+	"$(MAKE)" -C $(TDIR)/statconn
```
```suggestion
	"$(MAKE)" -C $(TDIR)/statconn -f $(RIOTBASE)/Makefile.base MODULE=nimble_statconn
```

And you can drop Makefile.nimble_statconn

aabadie commented:
> mkdir: cannot create directory ‘/home/mlenders/Repositories/RIOT-OS/RIOT/build/pkg’: Permission denied

Seems like an issue on your side.

miri64 commented:
@aabadie yepp. Github seems to not sync comments correctly at the time. See https://github.com/RIOT-OS/RIOT/pull/13552#issuecomment-651177137

miri64 commented:
> > But nodes `nrf52840dk-1-5` should be available, right? They should do the trick...
> 
> Ahhh sorry, I somehow scrolled over them. Sure! Will use them :-)

Sadly, all but nrf53840dk-3 are not flashable atm (always getting deployment status "Unknown") :(

haukepetersen commented:
welcome to my world :-)

fjmolinas commented:
Tested on 2 iotlab nrf52dk:

```
ifconfig
Iface  6  HWaddr: EE:60:D2:D5:D2:BD
          L2-PDU:1280 MTU:1280  HL:64  RTR
          6LO  IPHC
          Source address length: 6
          Link type: wireless
          inet6 addr: fe80::ee60:d2ff:fed5:d2bd  scope: link  VAL
          inet6 group: ff02::2
          inet6 group: ff02::1
          inet6 group: ff02::1:ffd5:d2bd
> statconn adds  C1:D4:80:83:11:CD
statconn adds  C1:D4:80:83:11:CD
success: connecting to peer as master
> ping6 fe80::c1d4:80ff:fe83:11cd
ping6 fe80::c1d4:80ff:fe83:11cd

--- fe80::c1d4:80ff:fe83:11cd PING statistics ---
3 packets transmitted, 0 packets received, 100% packet loss
> [ble] CONNECTED master (0|C1:D4:80:83:11:CD)
ping6 fe80::c1d4:80ff:fe83:11cd
ping6 fe80::c1d4:80ff:fe83:11cd
12 bytes from fe80::c1d4:80ff:fe83:11cd%6: icmp_seq=0 ttl=64 time=105.558 ms
12 bytes from fe80::c1d4:80ff:fe83:11cd%6: icmp_seq=1 ttl=64 time=80.547 ms
12 bytes from fe80::c1d4:80ff:fe83:11cd%6: icmp_seq=2 ttl=64 time=205.537 ms

--- fe80::c1d4:80ff:fe83:11cd PING statistics ---
3 packets transmitted, 3 packets received, 0% packet loss
round-trip min/avg/max = 80.547/130.547/205.537 ms
>
```

```
ifconfig
Iface  6  HWaddr: C1:D4:80:83:11:CD
          L2-PDU:1280 MTU:1280  HL:64  RTR
          6LO  IPHC
          Source address length: 6
          Link type: wireless
          inet6 addr: fe80::c1d4:80ff:fe83:11cd  scope: link  VAL
          inet6 group: ff02::2
          inet6 group: ff02::1
          inet6 group: ff02::1:ff83:11cd

> statconn addm  EE:60:D2:D5:D2:BD
statconn addm  EE:60:D2:D5:D2:BD
success: connecting to peer as slave
> [ble] CONNECTED slave (0|EE:60:D2:D5:D2:BD)
```

miri64 commented on pkg/nimble/statconn/include/nimble_statconn.h:
```diff
@@ -0,0 +1,171 @@
+ * @brief   Advertising interval used when in advertising mode [in ms]
+ */
+#ifndef NIMBLE_STATCONN_ADV_ITVL
+#define NIMBLE_STATCONN_ADV_ITVL            (90U)
```
Here and below: it became good practice to append the unit of time constats to the define name

```suggestion
#define NIMBLE_STATCONN_ADV_ITVL_MS         (90U)
```

Just a suggestion, not a must.

miri64 commented on pkg/nimble/statconn/nimble_statconn.c:
```diff
@@ -0,0 +1,267 @@
+
+static void _evt_dbg(const char *msg, int handle, const uint8_t *addr)
+{
+#if ENABLE_DEBUG
+    printf("[statconn] %s (%i|", msg, handle);
+    if (addr) {
+        bluetil_addr_print(addr);
+    }
+    else {
+        printf("n/a");
+    }
+    puts(")");
+#else
```
If you use C conditionals here, you can remove skip the `(void)var`s and avoid ugly code:

```suggestion
    if (ENABLE_DEBUG) {
        printf("[statconn] %s (%i|", msg, handle);
        if (addr) {
            bluetil_addr_print(addr);
        }
        else {
            printf("n/a");
        }
        puts(")");
    }
```

The optimizer should take care of the rest.

miri64 commented on sys/shell/commands/sc_nimble_netif.c:
```diff
@@ -29,7 +29,7 @@
 #include "net/bluetil/addr.h"
 
-#ifndef MODULE_NIMBLE_AUTOCONN
+#if !defined(MODULE_NIMBLE_AUTOCONN) && !defined(MODULE_NIMBLE_STATCONN)
```
```suggestion
#if !IS_USED(MODULE_NIMBLE_AUTOCONN) && !IS_USED(MODULE_NIMBLE_STATCONN)
```

miri64 commented on sys/shell/commands/sc_nimble_netif.c:
```diff
@@ -38,7 +38,7 @@
 #define DEFAULT_CONN_TIMEOUT        (500U)      /* 500ms */
 
-#ifndef MODULE_NIMBLE_AUTOCONN
+#if !defined(MODULE_NIMBLE_AUTOCONN) && !defined(MODULE_NIMBLE_STATCONN)
```
Same here and below

miri64 commented on sys/shell/commands/sc_nimble_netif.c:
```diff
@@ -378,7 +378,7 @@ void sc_nimble_netif_init(void)
 {
     if ((argc == 1) || _ishelp(argv[1])) {
-#ifndef MODULE_NIMBLE_AUTOCONN
+#if !defined(MODULE_NIMBLE_AUTOCONN) && !defined(MODULE_NIMBLE_STATCONN)
```
```suggestion
    if (!IS_USED(MODULE_NIMBLE_AUTOCONN) && !IS_USED(MODULE_NIMBLE_STATCONN) {
```

haukepetersen commented:
addressed comments:
- renamed the `NIMBLE_STATCONN_ADV_ITVL_MS` as proposed
- removed the debug output function: its not really needed anyway since I introduced the event callback
- using IS_USED in the shell command source file now

also rebased.

miri64 commented on pkg/nimble/statconn/include/nimble_statconn.h:
```diff
@@ -0,0 +1,171 @@
+ * @brief   Scan window used when opening new connections [in ms]
+ */
+#ifndef NIMBLE_STATCONN_CONN_WIN
+#define NIMBLE_STATCONN_CONN_WIN            (100U)
```
I meant this one too...

```suggestion
#define NIMBLE_STATCONN_CONN_WIN_MS         (100U)
```

miri64 commented on pkg/nimble/statconn/include/nimble_statconn.h:
```diff
@@ -0,0 +1,171 @@
+ * @brief   Opening a new connection is aborted after this time [in ms]
+ */
+#ifndef NIMBLE_STATCONN_CONN_TIMEOUT
+#define NIMBLE_STATCONN_CONN_TIMEOUT        (600U)
```
And this one

```suggestion
#define NIMBLE_STATCONN_CONN_TIMEOUT_MS     (600U)
```

miri64 commented on pkg/nimble/statconn/include/nimble_statconn.h:
```diff
@@ -0,0 +1,171 @@
+ * @brief   Connection interval used when opening a new connection [in ms]
+ */
+#ifndef NIMBLE_STATCONN_CONN_ITVL
+#define NIMBLE_STATCONN_CONN_ITVL           (75U)
```
```suggestion
#define NIMBLE_STATCONN_CONN_ITVL_MS        (75U)
```

miri64 commented on pkg/nimble/statconn/include/nimble_statconn.h:
```diff
@@ -0,0 +1,171 @@
+ * @brief   Supervision timeout used for new connections [in ms]
+ */
+#ifndef NIMBLE_STATCONN_CONN_SUPER_TO
+#define NIMBLE_STATCONN_CONN_SUPER_TO       (2500U)
```
```suggestion
#define NIMBLE_STATCONN_CONN_SUPER_TO_MS    (2500U)
```

haukepetersen commented:
ups, seems I did not think along... Also adapted the rest of the config params to carry the time unit in their name.

haukepetersen commented:
Nice, thanks for the review!
